### ISSUE TITLE
[ Question ] How to make a store instance per Component ?

### ISSUE THREAD

**Issue by a-eid** [react-spring/zustand#128]
let's say you have a `<Button />` component which some how stores it's state in `zustand` with `buttonStore` store.

now you want each button to have it's own state / store ( zustand store/state )

how would you achieve that ?

**Comment by a-eid** [comment]
I wonder if that's not valid question.

**Comment by zslabs** [comment]
Is this what you're after https://github.com/react-spring/zustand/issues/101

I had a similar question about passing down a store instance per Component to avoid prop-drilling, and it'd be a similar effect, but needing React context to handle the scope.

**Comment by dai-shi** [comment]
It's technically possible, but that may lead some pitfalls.
Creating store in React lifecycle can lead violating rule of hooks.

cc @drcmda This is another use case where splitting `useStore` and `create` makes sense.

**Comment by drcmda** [comment]
at work we are doing that

```jsx
const context = createContext()

function useStore(sel) {
  const useStore = useContext(context)
  return useStore(sel)
}

function Store({ children }) {
  const [useStore] = useState(() => create(...))
  return (
    <context.Provider value={useStore}>
      {children}
    </context.Provider>
  )
}

export { Store, useStore }
```

now every component inside Store can use the hook.

```jsx
import { Store, useStore } from "./store"

<Store>
  <Foo />
</Store>

function Foo() {
  const bar = useStore(state => state.bar)

```

**Comment by dai-shi** [comment]
```js
const [useStore] = useState(() => create(...))
```
Actually, this seems to work without problems.

But, in general passing a hook in props is not recommended.
```js
<Component useFoo={foo ? useFoo : useBar} />
```
if useFoo and useBar use different hook order, it will violate the rule.

In summary, if we carefully create stores, it will work. At least at the moment, yes.

Closing this. Feel free to reopen for further discussion.

**Comment by HarrisonB** [comment]
> In summary, if we carefully create stores, it will work. At least at the moment, yes.

Regarding this comment: does this mean that devs should not rely on this feature? Or just that they should be careful? Don't want to start adopting a pattern if a future update might break it.

**Comment by dai-shi** [comment]
@HarrisonB My comment was too conservative. Creating a store on demand is a valid pattern and it won't go away in the future.

```js
const [useStore] = useState(() => create(...))
```

I would actually prefer using `useRef` instead of `useState`, but it's not very handy because useRef doesn't have lazy initialization.

```js
const useStoreRef = useRef()
if (!useStoreRef.current) useStoreRef.current = create(...)
const useStore = useStoreRef.current
```

> in general passing a hook in props is not recommended.

This is still true. In other words, creating `useStore` and using it in a component / a hook is safe without caution.

**Comment by HarrisonB** [comment]
@dai-shi Thanks so much for the prompt response. I continue to love this library!

**Comment by merisbahti** [comment]
> @HarrisonB My comment was too conservative. Creating a store on demand is a valid pattern and it won't go away in the future.
> 
> ```js
> const [useStore] = useState(() => create(...))
> ```
> 
> I would actually prefer using `useRef` instead of `useState`, but it's not very handy because useRef doesn't have lazy initialization.
> 
> ```js
> const useStoreRef = useRef()
> if (!useStoreRef.current) useStoreRef.current = create(...)
> const useStore = useStoreRef.current
> ```
> 
> > in general passing a hook in props is not recommended.
> 
> This is still true. In other words, creating `useStore` and using it in a component / a hook is safe without caution.

Can we have this hook blessed by including it in utils or something? Thanks.

**Comment by dai-shi** [comment]
@merisbahti You want to check the discussion in #182. If what you are suggesting something different, would you open a new issue please?

**Comment by boltcoder** [comment]
Funny thing -> I was using react context providers and then was context bridging to work with react-three-fiber. Then I moved everything off from context and moved to zustand. Life was good. Now I need multiple instances of my 3d component. So I need to provide the zustand store I create via contextproviders bridged between three-fiber's canvas and normal react-dom. 
Thumbsup if you're all coming full circle like me 👯 

PS: I still love this library. You guys are great.

**Comment by gsouf** [comment]
I'm wondering what's the advantage of using 1 zustand instance for each component instance over using `useReducer`?

**Comment by dai-shi** [comment]
One use case is that we want to create a store in React tree, and then pass it with props or context. This is an old thread, and we now have `zustand/context` to cover such a use case.

**Comment by rachitmishra** [comment]
@dai-shi curios? instead of `zustand/context`;

 if we have something like
```
const useStoreScreenA = create((set) => ...)
const useStoreScreenB = create((set) => ...)
```
won't that also work? would this be affected by lifecycle or any pitfalls which i may be missing?

**Comment by dai-shi** [comment]
Yes, it works.
A pitfall is that passing a custom hook through React Context is not a good idea in general (as it can lead to violating rules of hooks).

`zustand/context` is a simple wrapper around `React.createContext` to avoid such a pitfall.

In zustand v4, we have
```js
import { createStore, useStore } from 'zustand'
```
so that you can safely use `React.createContext`. (and we will possibly deprecate `zustand/context` in the future.)

**Comment by wuarmin** [comment]
>> A pitfall is that passing a custom hook through >> React Context is not a good idea in general (as >> it can lead to violating rules of hooks).

Hey @dai-shi,
you mean, if we are passing a function, which returns a hook, we creating the possibility, that the receiver uses it wrong and violates the rules of hooks? So passing a function is recommended, but passing a higher order function, which returns  a hook is not recommend.

**Comment by dai-shi** [comment]
Yes.

**Comment by wuarmin** [comment]
Thanks!

**Comment by grantyi** [comment]
@dai-shi Would it possible to make a more robust example of using React Context + Zustand? I am having difficulty creating individual instances per component. When I create multiple instances of a component using React Context + Zustand, they seem to still be manipulating the same Zustand store

**Comment by dai-shi** [comment]
https://codesandbox.io/s/polished-pond-4jn1e2 is the only example I created.

> they seem to still be manipulating the same Zustand store

Sounds like there's some misusages??

**Comment by grantyi** [comment]
It would appear so! Thank you for sharing that example - I will have to see what I’m doing wrong

**Comment by YamabikoLab** [comment]
Thank you very much, @dai-shi . 
I like zuland very much.

I have one question about the following example. 

>https://codesandbox.io/s/polished-pond-4jn1e2 is the only example I created.

If I want to use "subscribe", how should I implement it?

**Comment by dai-shi** [comment]
Doesn't `store.subscribe(...)` work?

**Comment by YamabikoLab** [comment]
Thanks for the reply.
How can I implement a call to subscribe in the Counter component?

**Comment by YamabikoLab** [comment]
Thanks for the reply, @dai-shi .
How can I implement a call to subscribe in the Counter component?

**Comment by dai-shi** [comment]
It depends on how you would like to do it. Here's one example: https://codesandbox.io/s/stoic-blackwell-ndglrz?file=/src/App.tsx

**Comment by bhallstein** [comment]
> ```js
> const [useStore] = useState(() => create(...))
> ```
> I would actually prefer using `useRef` instead of `useState`, but it's not very handy because useRef doesn't have lazy initialization.

Glad to find this issue as it cleared this up for me. Just to add, this seems like a good case for [useMemo](https://reactjs.org/docs/hooks-reference.html#usememo).

**Comment by dai-shi** [comment]
@YamabikoLab How about this?
https://codesandbox.io/s/infallible-hoover-lbn7sm?file=/src/App.tsx

**Comment by YamabikoLab** [comment]
Thank you for your response, @dai-shi 
They were very helpful!
I will try to call subscribe with this implementation.

**Comment by Sejmou** [comment]
Thanks for everything that was already discussed in this thread, it's really helpful!
However, I do have one issue that has been puzzling me for a while: how do I make store selectors work with TypeScript?

E.g. in the `<Counter>` component from the code sandbox [example](https://codesandbox.io/s/polished-pond-4jn1e2) that was posted earlier, I want to be able to do

```typescript
const count = useMyStore(state => state.count);
const inc = useMyStore(state => state.inc);
```

instead of
```typescript
const { count, inc } = useMyStore();
```

[Here's](https://github.com/pmndrs/zustand/blob/main/docs/previous-versions/zustand-v3-create-context.md#migration) also an example for how to export the selector as well inside the `useMyStore()`, but the problem is that I don't understand what type annotation I need for its `selector` input arg (the example uses plain JS).

```javascript
export const useMyStore = (selector) => useStore(useContext(MyContext), selector); // what type do I need for the selector here?
```

**Comment by dai-shi** [comment]
This should do:
```ts
export const useMyStore = <T>(selector: (state: MyState) => T) => ...
```

I think we improved the docs recently.

**Comment by Sejmou** [comment]
Thank you for the quick response, that worked :)

**Comment by sbussard** [comment]
Do we all see the irony that zustand doesn't require context for global state but it does for local state?

**Comment by thimvanamersfoort** [comment]
Hi guys. For anyone still reading this now, I created the following TS code snippet to ease the usage of Zustand within a React Context:

```typescript
/**
 * Higher-order function to create a context provider and a custom hook for a given Zustand store.
 * 
 * @param createStore - A factory function that creates a new Zustand store.
 * @returns A tuple containing the context provider and the custom hook.
 */
export function createStoreContext<TState>(createStore: () => UseBoundStore<StoreApi<TState>>) {

  /**
   * React context created to provide the Zustand store to components. 
   * This context will hold the Zustand store created by `createStore`.
   */
  const StoreContext = createContext<UseBoundStore<StoreApi<TState>>>(createStore());

  /**
   * React component that provides the Zustand store to its children components. 
   * It uses the `createStore` function to create a store instance and provides it via `StoreContext`.
   */
  const StoreProvider:React.FC<{ children: ReactNode }> = ({ children }) => {
    const [ useStore ] = useState(createStore);
    return (
      <StoreContext.Provider value={useStore}>
        {children}
      </StoreContext.Provider>
    ) 
  };

  /**
   * Custom hook that provides access to the Zustand store within components. 
   * It uses `useContext` to access the store from `RowStoreContext` and returns the hook returned by the Zustand store.
   */
  const useStore = () => {
    const useStore = useContext(StoreContext);
    return useStore();
  };

  return [ StoreProvider, useStore ] as const;
}
```

Usage is as follows:

```typescript
// In typescript, optionally set the state type
type CounterStoreState = {
  counter: number,
  setCounter: (newNumber: number) => void
}

// Create the store you want to use
const createCounterStore = () => create<CounterStoreState>(
  (set, get) => ({
    counter: 0,
    setCounter: (newNumber) => {
      set({ counter: newNumber })
    }
  })
)

// First is the React Context provider to wrap your components in, second is the hook to directly access the store
const [ CounterStoreProvider, useCounterStore ] = createStoreContext(createCounterStore)

```

Hope this helps some of you out!

**Comment by janwirth** [comment]
@thimvanamersfoort how would I consume the store in a child component? There seems to be no way to access the context, so what is the point of context here? (I'm probably ignorant to some React techniques here).

**Comment by thimvanamersfoort** [comment]
@janwirth you can just use the `useStore` hook returned by the `createStoreContext` fn in your child component, like how you would use a normal Zustand store hook in your components. The state of the store will be shared by all components that are wrapped by the `StoreProvider` component that is also returned by `createStoreContext`.

The useStore hook returned by the function internally uses the native React useContext function, so it does have internal access to the context.

I don't exactly know your internal use-case, but for component specific shared state, [jotai](https://jotai.org/) is also a good alternative. Created by the same guys that created Zustand, but it can be scoped to specific components (and smaller scale contexts) more easily.

Hope this helps you.

**Comment by gsouf** [comment]
@janwirth this guide illustrate how to use a context with zustand https://docs.pmnd.rs/zustand/guides/initialize-state-with-props

Note that there is a mistake in this guide and you will want to make sure the store in the ref is created only once, so you must update the `useRef` part to be like that:

```
cons storeRef = useRef()
if (!storeRef.current) {
  storeRef.current = createStore(...)
}
```

**Comment by janwirth** [comment]
Thank you!

**Comment by robxbob** [comment]
Is there even a point in using a store per component if youre just going to use useContext? Wouldnt useContext cause the children components to always render per store change, even though the main goal of this store is to prevent rerenders when a state isnt being explicitly used in a child component?

**Comment by dai-shi** [comment]
If you create a stable store per component that's only created for the first time, child components don't always rerender.

**Comment by chrysb** [comment]
> @janwirth you can just use the `useStore` hook returned by the `createStoreContext` fn in your child component, like how you would use a normal Zustand store hook in your components. The state of the store will be shared by all components that are wrapped by the `StoreProvider` component that is also returned by `createStoreContext`.
> 
> The useStore hook returned by the function internally uses the native React useContext function, so it does have internal access to the context.
> 
> I don't exactly know your internal use-case, but for component specific shared state, [jotai](https://jotai.org/) is also a good alternative. Created by the same guys that created Zustand, but it can be scoped to specific components (and smaller scale contexts) more easily.
> 
> Hope this helps you.

Can you share an example of how a child component accesses the store?

Are you saying the child component would call `createStoreContext` again?

**Comment by Layhout** [comment]
I’m really thankful for this thread. it’s helped me a lot!
I do have a question though, what happens to the component’s store when the component is unmounted? Does the store still remain in memory even after the component is gone?

**Comment by dai-shi** [comment]
With:
```js
const storeRef = useRef(null)
if (storeRef.current === null) {
  storeRef.current = createStore(...)
}
```
The created store will be gone after unmounted.
